Add ufbx for FBX importing

fire opened the pull request:
This restores FBX imports in Godot.

I would need people to test their models, or downloaded models. Specifically, I want to test that:

* The model imports correctly, with textures and animation
* That the process is smooth and intuitive

To test this properly, a variety of models is required, so do not hesitate to try different ones. You can find free FBX models at https://sketchfab.com/3d-models?features=downloadable (be careful, you may see NSFW models).

If you run into an issue, please post here with your OS, and as much information as you can about the error.

https://github.com/godotengine/godot/assets/32321/e1af70c5-11bc-4c2a-9bad-6645cfb6963f

This work is supported as part of [V-Sekai](https://github.com/V-Sekai/) and with help from @bqqbarbhg.

The work was approved at the Animation / 3D Pipeline meeting and was also discussed with the production team as an experimental marked feature for 4.3.

1. https://godotengine.org/fbx-import/
2. https://godotengine.org/article/fbx-importer-rewritten-for-godot-3-2-4/
3. https://godotengine.org/article/here-comes-godot-3-2/#3d-assets

- [x] Marked experimental
- [x] Added ufbx to Godot Engine
- [x] Correct uv1 and uv2 for import.
- [x] Animation clips support
- [x] Skeletal animation support
- [x] Meshless fbx files are imported with animations
- [x] Blend shapes are imported
- [x] Allow Utf8 string name imports  
- [x] Postponed camera imports
- [x] Postponed light imports
- [x] Blend shape animations are imported
- [x] Eight uv support
- [x] Remove code scaffolding from GLTFDocument
- [x] Prune extra files in thirdparty/ufbx
- [x] Need help with the disable enable scheme with FBX2GLTF
- [x] Handle scale ignore [HARD]
- [x] Transparency is enabled
- [x] HANA bug with clusters being parented to the root and failing import
- [x] Postponed Camera imports again
- [x] Postponed Light imports again
- [x] Add documentation
- [x] Adapt FBXDocumentExtension

No proposal exists because this restores FBX import with around five to eight repeated attempts.

~~See also https://github.com/V-Sekai/godot-ufbx~~

### Finalization after 4.3 window is open

~~- [ ] Fix engine bugs on 2 gigabyte fbx samples (cc-by)~~
- [x] Move ufbx into the godot/thirdparty folder
- [x] Write licensing information

### Finalization before 4.3 window closes

1. [x] Cameras
2. [x] Lights
3. [x] ~~Curve paths~~

### Known Issues

When the fbx2gltf import has failed it will use the mostly recently cached scene from the ufbx import. 

See https://github.com/godotengine/godot/issues/86309 for the related issue.

_Production edit: closes godotengine/godot-roadmap#40_

bqqbarbhg commented:
By the way, ufbx also has optional support for importing Wavefront `.obj` files that is as exhaustively tested. I'm not sure what kind of support Godot has for `.obj` files currently but it might be worth comparing and considering.

fire commented:
@bqqbarbhg It's a good idea, but I would prefer to merge ufbx for FBX importing first and then we can expose the obj importer later. For clarity and maintenance ease.

fire commented:
The codebase is synced from https://github.com/V-Sekai/godot-ufbx and has commit logs.

DigitalN8m4r3 commented:
Tested with
*mixamo models
*rdyplayerme models
*custom models

so far no issues, good work 👍

fire commented:
From @bqqbarbhg the latest update supports fixes bone scaling when using unusual inherit modes.

> In addition to normal bones (RSrs, RS being rotation/scale matrix of parent, rs of the child), FBX supports bones that ignore the size of their parent (Rrs), or composite it into their own size (RrSs).
>
> Added a new feature in ufbx to create helper nodes to support these inherit modes in a standard scene graph/skeleton.

fire commented:
Some reports of error spam with images.

Also large files with animations error with:

```
ERROR: Condition "cache_update_size >= NODE_CACHE_UPDATE_MAX" is true. Continuing.
   at: AnimationPlayer::_animation_process_animation (scene\animation\animation_player.cpp:596)
ERROR: Condition "cache_update_size >= NODE_CACHE_UPDATE_MAX" is true. Continuing.
   at: AnimationPlayer::_animation_process_animation (scene\animation\animation_player.cpp:540)
ERROR: Condition "cache_update_size >= NODE_CACHE_UPDATE_MAX" is true. Continuing.
   at: AnimationPlayer::_animation_process_animation (scene\animation\animation_player.cpp:568)
ERROR: Condition "cache_update_size >= NODE_CACHE_UPDATE_MAX" is true. Continuing.
   at: AnimationPlayer::_animation_process_animation (scene\animation\animation_player.cpp:596)
```

sketchfab/PaulYang/bomb-crack-ground-00-free.fbx

sketchfab/Rigsters/deer-group-snow-globe.fbx

AThousandShips commented on modules/fbx/fbx_document.cpp:
```diff
@@ -0,0 +1,2809 @@
+
+		if (fbx_node->skeleton >= 0) {
+			const Skeleton3D *sk = p_state->skeletons[fbx_node->skeleton]->godot_skeleton;
+			ERR_FAIL_COND(sk == nullptr);
```
```suggestion
			ERR_FAIL_NULL(sk);
```

AThousandShips commented on modules/fbx/fbx_document.cpp:
```diff
@@ -0,0 +1,3167 @@
+	BoneAttachment3D *bone_attachment = memnew(BoneAttachment3D);
+	print_verbose("glTF: Creating bone attachment for: " + fbx_node->get_name());
+
+	ERR_FAIL_COND_V(!bone_node->joint, nullptr);
```
My bad misread this one

fire replied:
I resolved it. Thanks!

fire commented:
Needed to rebase because of https://github.com/godotengine/godot/pull/81968/files preventing me from testing mac.

fire commented:
TL;DR we bias the importer towards Maya fbx because it is the pipeline lacking representation in Godot Engine, 

Blender has a functional GLTF pipeline and if needed Blender has a workaround using the FBX exporter option **"FBX Units Scale"**.

## In-depth summary

Here's an ai summary of our conversation why to bias Maya FBX vs Blender FBX:

In the context of the godot-ufbx, we bias towards Maya FBX files over Blender FBX files by using `UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY` avoiding `UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS`. The `UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY` option in the ufbx library is used to scale the geometry of a model. This is similar to how the fbx2gltf tool operates. It's particularly useful for handling files that do not have the 100x scaling in nodes, which is common in many models exported from Maya.

When you use `UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY`, the 100x scale factor applied by Blender during export will be retained. This means that if you import a model into Blender that was exported with this option enabled, the model will appear correct but scaled.

If you are exporting your models from Blender, you can select the "FBX Units Scale" option before exporting. This will ensure that the scale factor is correctly set during export, avoiding the need for manual adjustment when importing back into Blender.

## Example

Export the test scene from Blender will save the file as:

```
Node (scale 100)
  Vertices [(1, 0, 0), (-1, 0, 0), (0, 1, 0)]
```

importing with UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS modifies it to

```
Node (scale 1)
  Vertices [(1, 0, 0), (-1, 0, 0), (0, 1, 0)]
```

and UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY modifies it to

```
Node (scale 100)
  Vertices [(0.01, 0, 0), (-0.01, 0, 0), (0, 0.01, 0)]
```

and on the other hand exporting a triangle like that from Maya would usually be like

```
Node_Maya (scale 1)
  Vertices [(100, 0, 0), (-100, 0, 0), (0, 100, 0)]

UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS
Node_Maya (scale 0.01)
  Vertices [(100, 0, 0), (-100, 0, 0), (0, 100, 0)]

UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY
Node_Maya (scale 1)
  Vertices [(1, 0, 0), (-1, 0, 0), (0, 1, 0)]
```

[blender_default.fbx.zip](https://github.com/godotengine/godot/files/12838626/blender_default.fbx.zip)

fire commented on modules/fbx/structures/fbx_skeleton.h:
```diff
@@ -0,0 +1,103 @@
+
+	Skeleton3D *get_godot_skeleton();
+
+	// Skeleton *get_godot_skeleton() {
+	// 	return this->godot_skeleton;
+	// }
+	// void set_godot_skeleton(Skeleton p_*godot_skeleton) {
+	// 	this->godot_skeleton = p_godot_skeleton;
+	// }
```
Remove unused code.

fire commented on modules/fbx/doc_classes/FBXDocument.xml:
```diff
@@ -0,0 +1,51 @@
+			<param index="1" name="base_path" type="String" />
+			<param index="2" name="state" type="FBXState" />
+			<param index="3" name="flags" type="int" default="0" />
+			<description>
+			</description>
```
Fill in all the documentation

fire commented on modules/fbx/doc_classes/FBXDocument.xml:
```diff
@@ -0,0 +1,51 @@
+			<description>
+			</description>
+		</method>
+		<method name="register_fbx_document_extension" qualifiers="static">
+			<return type="void" />
+			<param index="0" name="extension" type="FBXDocumentExtension" />
+			<param index="1" name="first_priority" type="bool" default="false" />
```
Reminder to fix.

fire commented on modules/fbx/thirdparty/ufbx/.github/workflows/ci.yml:
```diff
@@ -0,0 +1,556 @@
+name: CI
```
Can cleanup ufbx third party and update the readme citing the library.

fire commented on modules/fbx/doc_classes/EditorSceneFormatImporterUFBX.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
@AThousandShips if you're interested, you can take a look at the documentation. It's very rough.

adamscott commented on modules/fbx/README.md:
There's no README.md currently for modules, but it could change. There's a [PR for that in the GDScript module](https://github.com/godotengine/godot/pull/81345) that is on it's way to be merged.

But the difference between that PR and this PR's README.md is the fact that the former README.md is for new users in order to understand better on how the module works.

Right now, the current README.md is only displaying the `.all-contributorsrc` info, which is not done usually in the Godot project right now.

So could you remove this file? Thanks!

fire commented on modules/fbx/editor/editor_scene_importer_fbx.cpp:
```diff
@@ -0,0 +1,85 @@
+
+#ifdef TOOLS_ENABLED
+
+#include "../fbx_document.h"
```
Use a project path rather than a relative path.

reduz reviewed (approved):
Looks good on my side, this is really up to how much it was tested.

fire commented on thirdparty/README.md:
```diff
@@ -792,6 +792,15 @@ Files extracted from upstream source:
+
+Files extracted from upstream source:
+
+- `ufbx.{cc,h}`
```
```suggestion
- `ufbx.{c,h}`
```

fire commented on modules/fbx/fbx_document.cpp:
```diff
@@ -0,0 +1,2809 @@
+	return uint32_t(p_vertex.x) | uint32_t(p_vertex.y) << 16;
+}
+
+// static Vector3 _arr_to_vec3(const Array &p_array) {
+// 	ERR_FAIL_COND_V(p_array.size() != 3, Vector3());
+// 	return Vector3(p_array[0], p_array[1], p_array[2]);
+// }
+
+// static Quaternion _arr_to_quaternion(const Array &p_array) {
+// 	ERR_FAIL_COND_V(p_array.size() != 4, Quaternion());
+// 	return Quaternion(p_array[0], p_array[1], p_array[2], p_array[3]);
+// }
+
+// static Transform3D _arr_to_xform(const Array &p_array) {
+// 	ERR_FAIL_COND_V(p_array.size() != 16, Transform3D());
+
+// 	Transform3D xform;
+// 	xform.basis.set_column(Vector3::AXIS_X, Vector3(p_array[0], p_array[1], p_array[2]));
+// 	xform.basis.set_column(Vector3::AXIS_Y, Vector3(p_array[4], p_array[5], p_array[6]));
+// 	xform.basis.set_column(Vector3::AXIS_Z, Vector3(p_array[8], p_array[9], p_array[10]));
+// 	xform.set_origin(Vector3(p_array[12], p_array[13], p_array[14]));
+
+// 	return xform;
+// }
+
```
```suggestion
```

fire commented on modules/fbx/fbx_document.cpp:
```diff
@@ -0,0 +1,2809 @@
+	return OK;
+}
+
+// static Vector<uint8_t> _parse_base64_uri(const String &p_uri) {
+// 	int start = p_uri.find(",");
+// 	ERR_FAIL_COND_V(start == -1, Vector<uint8_t>());
+
+// 	CharString substr = p_uri.substr(start + 1).ascii();
+
+// 	int strlen = substr.length();
+
+// 	Vector<uint8_t> buf;
+// 	buf.resize(strlen / 4 * 3 + 1 + 1);
+
+// 	size_t len = 0;
+// 	ERR_FAIL_COND_V(CryptoCore::b64_decode(buf.ptrw(), buf.size(), &len, (unsigned char *)substr.get_data(), strlen) != OK, Vector<uint8_t>());
+
+// 	buf.resize(len);
+
+// 	return buf;
+// }
+
```
```suggestion
```

AThousandShips reviewed (approved):
Documentation style and wording LGTM

adamscott reviewed (changes requested):
Don't forget to rebase your PR.

AThousandShips commented:
Should this also be added to `COPYRIGHT.txt`?

adamscott commented:
> Should this also be added to `COPYRIGHT.txt`?

Yes. See @RandomShaper's commit (https://github.com/godotengine/godot/commit/fe4850c0d0e8eed3fe851007c667206684aab0fc) about mingw-std-threads.

bqqbarbhg commented:
> Should this also be added to `COPYRIGHT.txt`?

On my side, ufbx is dual licensed as public domain / MIT so it's not strictly necessary, but I do appreciate it if you want to credit it :D

adamscott commented:
> On my side, ufbx is dual licensed as public domain / MIT so it's not strictly necessary, but I do appreciate it if you want to credit it :D

We'll use the MIT license gladly! ;) Thanks for your work. <3

bqqbarbhg commented:
Related to the naming, would need to decide between `UFBX_GEOMETRY_TRANSFORM_HANDLING_HELPER_NODES` and `UFBX_GEOMETRY_TRANSFORM_HANDLING_MODIFY_GEOMETRY`. The current `UFBX_GEOMETRY_TRANSFORM_HANDLING_HELPER_NODES` always adds the dummy nodes in presence of geometry transforms (ie. 3ds Max pivots and such). `UFBX_GEOMETRY_TRANSFORM_HANDLING_MODIFY_GEOMETRY` on the other hand attempts to modify the geometry of meshes in cases where it is possible to avoid creating the helper nodes.

bqqbarbhg commented:
@fire Yeah in that case using helper nodes always explicitly preserves them. I guess it could also be added as an importer option if deemed necessary. In ufbx I have extensive tests that both options should have identical results though, so it only matters for the hierarchy.

fire commented:
Can you review the examples I sent you and let me know if this is feasible. It was one of the features that one of my peers wanted the most.

bqqbarbhg commented:
Yeah looks like it's not using geometry transforms but rather the `RotationPivot`/`ScalingPivot` feature. I have not yet implemented pivot conversion to ufbx but it's high on my TODOs, so I can try to implement it in ufbx.

fire commented:
1. Adding RotationPivot/ScalingPivot has high estimated complexity.
2. I don't want to delay the merge process and we can start working on the feature in a separate branch.
3. Merging first allows us to get the basic version out quickly, gather feedback, and then improve upon it in subsequent iterations.
4. Once the merge is complete, we can focus all our efforts on implementing the rotation/scaling pivot conversion.

We also discussed before shipping needing to think about how to allow users to migrate as well, ideally old projects should keep using fbx2gltf at least for some versions.

My strategy is to opt into fbx2gltf because it's a third party binary so we can import using ufbx as a first party feature and then if people configure fbx2gltf they can use that. However fbx2gltf/ufbx is a global setting which might not be ideal.

fire commented:
How to handle the transition between these fbx2gltf?

## Alternative 1: Per-Project or Per-Asset Setting

In this approach, we would create a setting that allows users to choose the tool they want to use based on their specific project or asset requirements. If the project is old, we would set the tool to `gltf2fbx`. For newer projects, we would set it to use `ufbx`.

## Alternative 2: Improved Import Logic

In this alternative, we would enhance the import logic. The system would default to using `ufbx` unless the file was previously imported with `fbx2gltf`. In such cases, the import would be blocked.

aaronfranke commented:
@fire @adamscott What is gltf2fbx? Do you mean FBX2glTF?

fire commented:
I and Reduz discussed that we prefer a per asset setting with `ufbx` as the default option and the fbx2gltf tool is used as a fallback in cases where a specific asset might not function properly and that a bug fix is still pending.

fire commented:
@reduz Do you know how that new system we added for compatibility breakage works?

aaronfranke commented:
@fire You override the `handle_compatibility_options` method.

fire commented:
Any suggestions?

1. When I import a ufbx it imports fine
2. When I then import a fbx2gltf it should fail
3. What happens is fbx2gltf loads the ufbx scn
4. !?

Any ideas how to fix?

J-Ponzo commented:
I successfully imported the 270 in-place animations from this pack https://assetstore.unity.com/packages/3d/animations/ultimate-traversal-anims-217762 (visualized one by one and compared with unity version when in doubt).

Note I could not check the root motion versions because they are not imported as an animated skeleton, but as an animated hierarchy of Node3D. I think this is because for some reason, the author designed the root motion versions differently (when I open the files in blender, they consist of a hierarchy of animated empties instead of the classic armature/bones design).

It could make sense that empties are interpreted as Node3D by the importer. But since the "animated empty" pattern is very common for animation packs, it would be great to have an import option giving a hint to the importer so that he knows whether it needs to interpret empties as skeleton bones or not. But for me it's more a feature request than an actual issue.

If you think this is useful, I can test with other animation packs and report the results here.

adamscott commented:
> 3\. What happens is fbx2gltf loads the ufbx scn

When does this happen?

fire commented:
My friend requested I test splines and so I checked that it fails as expected.

![image](https://github.com/godotengine/godot/assets/32321/1afde7b0-bdff-4d39-aca6-06b6a759bbf0)

![image_02](https://github.com/godotengine/godot/assets/32321/d874fc54-5491-4350-9c58-e64f9216b22e)
![image](https://github.com/godotengine/godot/assets/32321/f0dd709f-8219-4e1c-b68c-9b8cc90c0ff2)


[FBX Game Project.zip](https://github.com/godotengine/godot/files/13718725/FBX.Game.Project.zip)

fire commented:
> It could make sense that empties are interpreted as Node3D by the importer. But since the "animated empty" pattern is very common for animation packs, it would be great to have an import option giving a hint to the importer so that he knows whether it needs to interpret empties as skeleton bones or not. But for me it's more a feature request than an actual issue.

@lyuma is proposing a method where we can arbitrarily say a section of the node tree is a skeleton, or a node or cannot be a skeleton bone, or cannot be a node.

J-Ponzo commented:
> @lyuma is proposing a method where we can arbitrarily say a section of the node tree is a skeleton bone, or a node or cannot be a skeleton bone, or cannot be a node.

Sounds great! I can't wait to test that.

fire commented:
Thanks to work by @bqqbarbhg to merge the three PRs that have been hanging the ufbx in the PR is now from this commit: https://github.com/ufbx/ufbx/tree/85ed46443798d8824cb373451052ada6946e522e

This also introduces threads to speed up the ufbx portion of loading.

fire commented:
I spent some time testing this pr and thought it would be best to ask for feedback on 
![image](https://github.com/godotengine/godot/assets/32321/de80859d-72ed-4113-b3b6-45253e1ccd41) wording

fire commented:
Wording changed from

> "FBX2glTF is required for importing FBX files.\nPlease download it and provide a valid path to the binary:"

to

> "FBX2glTF is required for importing FBX files if using FBX2glTF.\nAlternatively, you can use ufbx by disabling FBX2glTF.\nPlease download the necessary tool and provide a valid path to the binary:"

akien-mga commented:
I reviewed the code locally and tested briefly.

Here's a branch with a second commit that cleans things up a bit, some style fixes, buildsystem consistency refactoring (and notably only building `ufbx.c` requires passing custom configuration defines, and warnings should be disabled), and removing a bunch of dead code and unused headers copied from glTF.

https://github.com/akien-mga/godot/tree/vsk-ufbx
https://github.com/akien-mga/godot/commit/9b696613f966e6bb36bb6301ffc1d2d1e8bad14f

You can amend these changes in your commit directly, or I can do it, if you agree with the changes.

---

Then I tested the feature briefly with https://polyhaven.com/a/chess_set (FBX, 1k), and it doesn't seem to handle textures properly for me.

In Godot 4.2.1 with FBX2glTF, it works ok-ish, albeit with a backface culling issue or similar:
![image](https://github.com/godotengine/godot/assets/4701338/3960fd08-8f76-457d-b21d-6574a93b40bc)

In this PR with ufbx, it fails loading the textures, it seems to treat the local path (`textures/chess_set_pieces_white_diff_1k.png`) as absolute (`res://textures/chess_set_pieces_white_diff_1k.png`), which doesn't work as I put that fbx file and its `textures` subfolder in a folder named `chess_set_1k`.

I get those errors:
```
ERROR: Resource file not found: res://textures/chess_set_pieces_white_diff_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_white_diff_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '0' couldn't be loaded from path: textures/chess_set_pieces_white_diff_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_pieces_white_nor_gl_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_white_nor_gl_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '1' couldn't be loaded from path: textures/chess_set_pieces_white_nor_gl_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_pieces_white_rough_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_white_rough_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '2' couldn't be loaded from path: textures/chess_set_pieces_white_rough_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_board_diff_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_board_diff_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '3' couldn't be loaded from path: textures/chess_set_board_diff_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_board_nor_gl_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_board_nor_gl_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '4' couldn't be loaded from path: textures/chess_set_board_nor_gl_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_board_rough_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_board_rough_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '5' couldn't be loaded from path: textures/chess_set_board_rough_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_pieces_black_diff_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_black_diff_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '6' couldn't be loaded from path: textures/chess_set_pieces_black_diff_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_pieces_black_nor_gl_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_black_nor_gl_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '7' couldn't be loaded from path: textures/chess_set_pieces_black_nor_gl_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
ERROR: Resource file not found: res://textures/chess_set_pieces_black_rough_1k.png (expected type: )
   at: _load (./core/io/resource_loader.cpp:279)
ERROR: Can't open file from path 'textures/chess_set_pieces_black_rough_1k.png'.
   at: get_file_as_bytes (./core/io/file_access.cpp:748)
WARNING: FBX: Image index '8' couldn't be loaded from path: textures/chess_set_pieces_black_rough_1k.png because there was no data to load. Skipping it.
     at: _parse_images (modules/fbx/fbx_document.cpp:1009)
WARNING: Ignoring face with non-finite normal in LOD generation.
     at: generate_lods (./scene/resources/importer_mesh.cpp:518)
```

![image](https://github.com/godotengine/godot/assets/4701338/46fbe5d9-a209-4f5c-a5c8-01b4051ab46b)

---

Changing the importer type to FBX2glTF also doesn't seem to work, I still get the same errors and references to `modules/fbx` code.

fire commented:
Added Akien's review changes and also [Handle textures relative to the FBX file and not res://.](https://github.com/godotengine/godot/pull/81746/commits/76e0a66acad16d98369b35cd9646aa65cab8cb47)

akien-mga commented:
I confirm the latest changes for relative paths seem to work fine on my simple test. Should probably be tested on more exhaustive test suite with different types of textured models.

An in-depth review by @lyuma would be great, otherwise from my point of view this should be good to go.

> Changing the importer type to FBX2glTF also doesn't seem to work, I still get the same errors and references to modules/fbx code.

This is still a problem but not blocking IMO, this can be worked on later.

fire commented:
From the discussion in rocketchat. I added `fbx/allow_geometry_helper_nodes` which is defaulted to off.

Can test with fbx in. This is a 3dmax file that is affected by allow_geometry_helper_nodes. https://sketchfab.com/3d-models/sci-fi-luminaris-spaceship-fb68179b3f8f4c9c80b528f9a331cafd.

adamscott commented:
> "FBX2glTF is required for importing FBX files if using FBX2glTF.\nAlternatively, you can use ufbx by disabling FBX2glTF.\nPlease download the necessary tool and provide a valid path to the binary:"

I don't know if this is the right way. The UX seems bad at best.

I suggest that we detect first FBX2glTF. If it's not detected, then we can skip this whole ordeal.

fire commented:
Instead of first detecting fbx2glTF, I removed the fbx support query code and now it does ufbx directly and the option to use fbx2glTF is a menu option.

adamscott commented:
> Instead of first detecting fbx2glTF, I removed the fbx support query code and now it does ufbx directly and the option to use fbx2glTF is a menu option.

That's awesome. But there's a small bug, we can reimport using "fbx2glTF" without having it set.
![image](https://github.com/godotengine/godot/assets/270928/e8355870-70e7-4df7-be2a-3585c5c26755)

And (as a small nitpick), I would rename `fbx/importer_type` to `fbx/importer`, as it's not the type of the importer that changes, but the importers themselves.

lyuma commented:
There was a lot of concern on Rocket Chat about the duplicate data model classes (FBXSkeleton vs GLTFSkeleton).

It would make it far easier to give this PR a proper review once these classes were merged, so I would appreciate doing this process as part of the next step.

adamscott commented:
> It would make it far easier to give this PR a proper review once these classes were merged

What needs to be merged before this PR? Do you have a link?

fire commented:
From Reduz

> fire so in the end, may it not be better to just unify the GLTF and FBX documents in a single thing? Given they contain roughly the same data. something like ImportedAssetDocument

So estimated time is it'll probably take next week to work on it.

fire commented on editor/import/3d/post_import_plugin_skeleton_renamer.cpp:
```diff
@@ -52,14 +52,14 @@ void PostImportPluginSkeletonRenamer::_internal_process(InternalImportCategory p
-
-	// Rename bones in Skeleton3D.
-	{
+	if (skeleton) {
```
Need to move to a separate pr.

fire commented:
- [ ] Moved FBXDocument and FBXSTate to their GLTF classes
- [ ] FBXLights and FBXCameras still has duplication

fire commented:
I got stuck, reverting to last version that imports fbx and I'll try again.

fire commented on modules/fbx/doc_classes/FBXCamera.xml:
```diff
@@ -0,0 +1,21 @@
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="depth_far" type="float" setter="set_depth_far" getter="get_depth_far" default="4000.0">
```
- [ ] Make sure the properties are overriden from the base GLTFCamera class

fire commented on modules/fbx/doc_classes/FBXDocument.xml:
```diff
@@ -0,0 +1,29 @@
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="register_fbx_document_extension" qualifiers="static">
```
Use the upstream gltf extension

fire commented on modules/fbx/structures/fbx_camera.cpp:
```diff
@@ -0,0 +1,118 @@
+
+#include "scene/3d/camera_3d.h"
+
+void FBXCamera::_bind_methods() {
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "perspective"), "set_perspective", "get_perspective");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "fov"), "set_fov", "get_fov");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "size_mag"), "set_size_mag", "get_size_mag");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "depth_far"), "set_depth_far", "get_depth_far");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "depth_near"), "set_depth_near", "get_depth_near");
+}
```
```suggestion
void FBXCamera::_bind_methods() {
}
```

Remove

fire commented on scene/resources/skin_tool.h:
```diff
@@ -0,0 +1,104 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#pragma once
```
Convert to header guards

fire commented on core/templates/template_convert.h:
```diff
@@ -1,5 +1,5 @@
 /**************************************************************************/
-/*  gltf_template_convert.h                                               */
+/*  template_convert.h                                                    */
```
Need reviewer commentary.

fire commented on modules/fbx/config.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def get_doc_classes():
+    return [
+        "EditorSceneFormatImporterUFBX",
+        "FBXCamera",
+        "FBXDocument",
+        "FBXDocumentExtension",
+        "FBXLight",
+        "FBXMesh",
+        "FBXState",
+        "FBXTexture",
+    ]
+
```
```suggestion
    return [
        "EditorSceneFormatImporterUFBX",
        "FBXCamera",
        "FBXDocument",
        "FBXLight",
        "FBXState",
    ]

```

fire commented:
We discussed the structure of ufbx at chat.godotengine.org, specifically the inheritance hierarchy of `FBXDocument` and `GLTFDocument`. 

It was proposed that `FBXDocument` should inherit from `GLTFDocument`, but adamscott pointed out that it doesn't make sense feature-wise as every `FBXDocument` would then be a `GLTFDocument`. Fire mentioned that this approach was suggested to avoid type duplication, which was identified as a review blocker. Fire also suggested renaming `GLTFDocument` to `AssetDocument3D` and deprecating the old name. Adamscott agreed with the idea of extending `AssetDocument3D` without making significant changes.

Fire thinks that we should keep `FBXDocument` extends `GLTFDocument` as is for now since there it is more complicated to move the class. What are your thoughts @lyuma & @reduz.

fire commented on modules/gltf/doc_classes/AssetDocument3D.xml:
```diff
@@ -0,0 +1,108 @@
+			<return type="int" enum="Error" />
+			<param index="0" name="bytes" type="PackedByteArray" />
+			<param index="1" name="base_path" type="String" />
+			<param index="2" name="state" type="GLTFState" />
```
Move the GLTFState specific class to GLTFDocument and make a new append_from_buffer, etc [...]

fire commented:
I want to break out the GLTFDocument Extends AssetDocument3D to a separate pr:

1. make sure the change from recent GLTFDocument patches don't get lost.
2. Unify the exporting apis so they use AssetState3D instead of GLTFState
3. Make AssetDocument3D an interface with minimal concrete methods
4. Rebase this ufbx pr on AssetDocument3D

fire commented:
@AThousandShips Due to reviewer request, everything is rearranged, would like some help to review style and convention.

AThousandShips commented on scene/resources/skin_tool.cpp:
```diff
@@ -0,0 +1,785 @@
+
+	out_roots.sort();
+
+	ERR_FAIL_COND_V(out_roots.size() == 0, FAILED);
```
```suggestion
	ERR_FAIL_COND_V(out_roots.is_empty(), FAILED);
```

AThousandShips reviewed (commented):
Will take a look over the documentation and another pass of the code later but think these are what sticks out 🙂

fire commented on modules/fbx/structures/fbx_light.h:
```diff
@@ -0,0 +1,77 @@
+	static Ref<FBXLight> from_node(const Light3D *p_camera);
+	Light3D *to_node() const;
+
+	void set_local_direction(const Vector3 &p_local_direction) { local_direction = p_local_direction; }
```
@bqqbarbhg Any idea how to remove this local direction variable?

fire commented on modules/fbx/structures/fbx_light.h:
```diff
@@ -0,0 +1,77 @@
+	void set_type(int p_type) { type = p_type; }
+	void set_decay(int p_decay) { decay = p_decay; }
+	void set_area_shape(int p_area_shape) { area_shape = p_area_shape; }
+	void set_inner_angle(float p_inner_angle) { inner_angle = p_inner_angle; }
```
There are duplicated in the parent class.

fire commented on modules/fbx/doc_classes/EditorSceneFormatImporterUFBX.xml:
```diff
@@ -0,0 +1,11 @@
+		Import FBX files using the ufbx library.
+	</brief_description>
+	<description>
+		EditorSceneFormatImporterUFBX is designed to load FBX files and supports both binary and ASCII FBX files from version 3000 onward. This class handles invalid files and out of memory errors gracefully, and supports various 3D object types like meshes, skins, blend shapes, materials, and rigging information. It provides utilities for evaluating loaded scenes, such as polygon triangulation, index  ...
```
```suggestion
		EditorSceneFormatImporterUFBX is designed to load FBX files and supports both binary and ASCII FBX files from version 3000 onward. This class supports various 3D object types like meshes, skins, blend shapes, materials, and rigging information. The class aims for feature parity with the official FBX SDK and supports FBX 7.4 specifications.
```

AThousandShips commented on modules/gltf/editor/editor_scene_importer_gltf.h:
```diff
@@ -50,6 +50,14 @@ class EditorSceneFormatImporterGLTF : public EditorSceneFormatImporter {
 			List<ResourceImporter::ImportOption> *r_options) override;
 	virtual void handle_compatibility_options(HashMap<StringName, Variant> &p_import_params) const override;
+	virtual Variant get_option_visibility(const String &p_path, bool p_for_animation,
+			const String &p_option, const HashMap<StringName, Variant> &p_options) {
```
```suggestion
			const String &p_option, const HashMap<StringName, Variant> &p_options) override {
```

AThousandShips reviewed (approved):
Went back over and LGTM

fire commented:
We found that certain models lose its fbx bind pose when I reset the skeleton bqqbarbhg assisted in debugging this.

lyuma commented:
I'm having trouble with image importing. It seems to be making duplicate copies of the referenced textures next to the FBX (though it uses the incorrect path fbx files with . in the name)

What I want is for an FBX which references textures to reference those copies, not create duplicates next to the fbx file.

Also small nitpick but the old `Error append_from_file` function does not need to be virtual. I do think it is possible to migrate these functions into the common base class without compatibility breakage without needing to rename them. This is how we got methods from AnimationPlayer migrated to AnimationMixer.

fire commented:
Can you explain which Error append_from_file?

fire commented:
Can you find a recreation project? for `I'm having trouble with image importing. It seems to be making duplicate copies of the referenced textures next to the FBX (though it uses the incorrect path fbx files with . in the name)`

lyuma commented:
Ok I figured out the cause: The fbx files had absolute paths hardcoded in them... to files located on my own PC.

In this case, godot seems to be making copies of those files from elsewhere on my hard disk, but with numbers like _1, _2, _3 etc.
If the referenced textures are actually in the same directory, then Godot uses them as is without making duplicates.

I don't think this flow works reliably and it's certainly surprising. For example, I would be unable to share my sample project with you because you would not have the same folders on your hard disk as me, so the fbx would import as all white.

So my recommendation is to ignore the absolute path component, or if you want to respect it, make sure to copy it into a project in a way that is findable in the future, not using numbers. Perhaps it should go in a textures folder.

fire commented:
The godot convention is putting the files next to the `.fbx` or `.glb`, but the names could be better.

fire commented:
Added a localize:

```c++
String path = _as_string(fbx_texture_file.filename);
path = ProjectSettings::get_singleton()->localize_path(path);
```

fire commented:
- [ ] nodes and meshes share the same unique name pool. this is causing the node to be renamed.
- [ ] When going through the embedded textures codepath do not use absolute paths

fire commented:
@lyuma 

```patch
% git diff
diff --git a/modules/fbx/fbx_document.cpp b/modules/fbx/fbx_document.cpp
index fa248d3306..3f049414e9 100644
--- a/modules/fbx/fbx_document.cpp
+++ b/modules/fbx/fbx_document.cpp
@@ -391,7 +391,8 @@ Error FBXDocument::_parse_meshes(Ref<FBXState> p_state) {
                if (fbx_mesh->name.length > 0) {
                        mesh_name = _as_string(fbx_mesh->name);
                }
-               import_mesh->set_name(_gen_unique_name(p_state->unique_names, mesh_name));
+
+               import_mesh->set_name(_gen_unique_name(p_state->unique_names, vformat("%s_%s", p_state->scene_name, mesh_name)));
 
                bool use_blend_shapes = false;
                if (fbx_mesh->blend_deformers.count > 0) {
```

fire commented:
> When going through the embedded textures codepath do not use absolute paths

Regarding this, I need a test file because logically I don't see any paths that avoid the absolute path.

lyuma commented:
```diff
diff --git a/modules/fbx/fbx_document.cpp b/modules/fbx/fbx_document.cpp
index fa248d3306..3e1764d052 100644
--- a/modules/fbx/fbx_document.cpp
+++ b/modules/fbx/fbx_document.cpp
@@ -391,7 +391,7 @@ Error FBXDocument::_parse_meshes(Ref<FBXState> p_state) {
 		if (fbx_mesh->name.length > 0) {
 			mesh_name = _as_string(fbx_mesh->name);
 		}
-		import_mesh->set_name(_gen_unique_name(p_state->unique_names, mesh_name));
+		import_mesh->set_name(_gen_unique_name(p_state->unique_mesh_names, mesh_name));
 
 		bool use_blend_shapes = false;
 		if (fbx_mesh->blend_deformers.count > 0) {
diff --git a/modules/fbx/fbx_state.h b/modules/fbx/fbx_state.h
index e62457d6d4..3daa334275 100644
--- a/modules/fbx/fbx_state.h
+++ b/modules/fbx/fbx_state.h
@@ -59,6 +59,7 @@ class FBXState : public GLTFState {
 	Vector<GLTFSkinIndex> skin_indices;
 	HashMap<ObjectID, GLTFSkeletonIndex> skeleton3d_to_fbx_skeleton;
 	HashMap<ObjectID, HashMap<ObjectID, GLTFSkinIndex>> skin_and_skeleton3d_to_fbx_skin;
+	HashSet<String> unique_mesh_names; // Not in GLTFState because GLTFState prefixes mesh names with the scene name (or _)
 
 protected:
 	static void _bind_methods();
```

fire commented:
Applied renamings from the meeting.

```
AssetDocument3D -> ModelDocument3D
AssetState3D -> ModelState3D
```

fire committed:
Add ufbx for FBX importing

This update introduces a new import method for FBX files using ufbx. If the fbx2gltf import fails, it will use the most recently cached scene from the ufbx import. The process is sped up by introducing threads to load the ufbx portion.

Key changes include:

- Support for importing geometry helper nodes in FBX files.
- Addition of cameras and lights with updated names.
- Removal of the fbx importer manager.
- Introduction of ModelDocument3D and updates to its methods.
- Changes to FBX import options and visibility.
- Updating the documentation and handling some errors.
- Store the original non-unique node, mesh and animation names in FBX and glTF.

Co-Authored-By: bqqbarbhg <bqqbarbhg@gmail.com>

fire commented:
Add "Store the original non-unique node, mesh and animation names in FBX and glTF."

fire commented:
Trying to fix the last remaining absolute path bug for now at https://github.com/V-Sekai/godot/pull/29

fire commented:
Merged in https://github.com/V-Sekai/godot/pull/30

fire commented:
https://github.com/ufbx/ufbx/tree/v0.11.0 is now tagged from BQQ

lyuma commented on scene/resources/skin_tool.h:
```diff
@@ -0,0 +1,101 @@
+using SkinNodeIndex = int;
+using SkinSkeletonIndex = int;
+
+class SkinTool : public Resource {
+	GDCLASS(SkinTool, Resource);
```
SkinTool consists only of static functions and is not exposed to the user, as it is a shared implementation detail of the GLTF and FBX importers. Because it will never be instantiated, it does not need to extend Resource (or Object) and does not require GDCLASS.

```suggestion
class SkinTool {
```

lyuma replied:
Namespace could also work, but namespaces don't have `private:` so I would prefer to keep it a class with static functions

fire replied:
I can put it inside of the ModelDocument3D header but that seems awkward.

lyuma replied:
Not a bad idea... ModelDocument3D seems just as good a place as any: it is a set of utility functions for model document implementations

lyuma reviewed (approved):
The code is excellent, and ufbx is an extremely capable library that seems to handle everything I've thrown at it so far. All of my comments are renames or removal of dead code.

The only open question I have is if there is a way to avoid renaming `generate_scene` to `create_scene`. I am not convinced a rename is necessary to keep compatibility.

Great job you two! LGTM

lyuma commented:
@fire it seems that through one of the merges or rebases, I was assigned as the commit author, but my changes were not substantive enough to claim that title.

I would suggest performing a --reset-author (and make sure that bqq remains as Co-Authored-By). I wasn't expecting to be a co-author

(Also feel free to update to include a minor upstream change in ufbx)

fire commented on modules/gltf/doc_classes/ModelDocument3D.xml:
```diff
@@ -0,0 +1,59 @@
+			<description>
+			</description>
+		</method>
+		<method name="create_scene">
+			<return type="Node" />
```
It needs to be a new name, I am not attached to `create_scene` but it cannot be `generate_scene`.

fire replied:
Could be `create_scene_from_data` to match the pattern.

fire replied:
`virtual Node *generate_scene_from_data(Ref<ModelState3D> p_state, float p_bake_fps = 30.0f, bool p_trimming = false, bool p_remove_immutable_tracks = true) override;`

vs

`Node *GLTFDocument::generate_scene_81746(Ref<GLTFState> p_state, float p_bake_fps, bool p_trimming, bool p_remove_immutable_tracks) {
`

Note `GLTFState` extends `ModelState3D`

lyuma replied:
If we have to change it, I prefer create_scene like you had it.

but I am not satisfied with the answer you gave. Why can't we call it generate_scene given the existence of generate_scene_81746

fire replied:
We can call it generate_scene but it will be expose to gdscript as `generate_scene_from_data` since it's a pattern in other parts of godot engine that we change name rather than change the parameter types.

lyuma replied:
But we haven't changed any of the parameter types, have we?

fire replied:
Ref<GLTFState> p_state became Ref<ModelState3D> p_state

lyuma replied:
Two things:
1. A Ref<ModelState3D> parameter can cleanly accept a Ref<GLTFState> with no conversion loss. Therefore this change should not be compatiblity breaking, certainly not from the perspective of GDScript.

As for GDExtension, the bind_compat_method should take care of the compatibility.

2. FBXState is a subclass of GLTFState, so there is no need to change this argument type yet. We can delay using the correct ModelState3D type until whatever support for contravariance is missing from point 1 is implemented.

in conclusion, I still see no reason to require a new method name.

And I am fairly sure the same applies to the other methods such as append_data_...

fire replied:
Applied. Please check.

fire commented:
Updated ufbx to [v0.11.1](https://github.com/ufbx/ufbx/commit/e3dcaa4512ddd549dd421190485c667ef19a0756)

fire commented:
@dsnopek Can you check if this is the correct way to do the gdextension changes?

fire commented:
> blend shape animation code has broken during the FBX/GLTF refactor (probably due to FBX requiring to store the blend shape tracks separately)

Discovered by @bqqbarbhg

fire commented:
Godot Engine supports visibility animation, and we can use the support in FBX for setting visibility, this in contrast to gltf2 which doesn't support visibility in the base specification.

However, we're deciding to fix the blend shape regression first.

fire commented:
We found some animation trimming issues.

For example on the shibahu model with trimming it looks good:

![image](https://github.com/godotengine/godot/assets/32321/2e743231-df12-4ebf-a660-2b0224e6257d)

but without trimming the keys are offset from the start (as they should be) but the animation time ends before the animation ends

![image](https://github.com/godotengine/godot/assets/32321/3ce22776-9c48-4a30-b002-d3a47aa6f51e)

so @bqqbarbhg recommends something like this:

```c++
double time_begin = additional_animation_data["time_begin"];
double time_end = additional_animation_data["time_end"];
double length = p_trimming ? time_end - time_begin : time_end;
animation->set_length(length);
```

fire commented:
We discussed how FBX trimmed would be a better default for FBX given that people often have animations that don't start from 0 :D but I understand not wanting to change 3d model import defaults.

Bromeon commented on misc/extension_api_validation/4.2-stable.expected:
```diff
@@ -147,3 +147,15 @@ GH-84792
 Added fog mode argument. Compatibility method registered.
+
+
+GH-81746
+--------
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/append_from_buffer
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/append_from_file
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/append_from_scene
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/generate_buffer
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/generate_scene
+Validate extension JSON: API was removed: classes/GLTFDocument/methods/write_to_filesystem
```
These are breaking changes in GDExtension, and afaik `GLTFDocument` is a stable (not experimental) API.

fire replied:
@lyuma I will revert to the changes earlier then unless you can submit a pr to change it.

fire commented:
If someone wants to do the append_from_buffer, append_from_file, append_from_scene, generate_buffer, generate_scene, write_to_filesystem compatibility changes please submit them as a pr to this branch on V-Sekai. I am not doing that for now as I renamed the methods in the base class of ModelDocument3D.

I'll nudge this pr to merged sooner than later.

fire commented:
I would like to reiterate that we should adhere to the consensus of merging the FBX PR as it is. Given the substantial amount of work involved, we could easily get lost in trying to fix bugs.

It has been suggested that we create pull requests against this particular pull request.

akien-mga commented:
I made another clean up pass in 5c2b0a0bbf0aad6c1919fb5ce75849efc48509f9, please review **and test** the changes (I didn't test, only made sure it compiles with different options).

I'm really not satisfied with the way gltf (and even fbx) includes found their way to the core and create a tight coupling that shouldn't exist.

The proper architecture should be `scene/3d/model_*.h` stuff as the abstraction for GLTF and FBX models. These **CANNOT** depend on any code in the gltf and fbx modules! Otherwise they're not abstractions at all. And that's a major problem currently so I had to move `scene/3d/model_*.h` back to the gltf module, after removing its coupling to the fbx module...

If that's not possible, then the second best would be to have the fbx module depend fully on gltf (which it does in the current PR, despite the attempt at abstracting this away), reusing as much code as possible, instead of duplicating it like the original version of this PR did.

I wish these aspects had been properly discussed and designed before brute-forcing an implementation that leads to significant coupling and tech debt. There's still a lot of code duplication and unclear responsibilities in the current version of the PR.

After spending another couple hours doing a second clean up pass, I'm also willing to give it a rest. So I echo the consensus to merge as is (if tests confirm it works as intended).
I'm not satisfied with the outcome on an architectural point of view, so I hope someone does a thorough re-design of this architecture in a way that matches the standards we strive for usually for adding new features. But this can be done in the future, probably.

akien-mga commented on modules/fbx/config.py:
```diff
@@ -1,4 +1,5 @@
 def can_build(env, platform):
+    env.module_add_dependencies("fbx", ["gltf"])
```
This makes the tight coupling between fbx and gltf explicit. `fbx` can't be compiled without `gltf` (but `gltf` can be compiled without `fbx`, after I moved FBX2glTF here due to the introduced coupling).

akien-mga commented on modules/fbx/editor/editor_scene_importer_ufbx.cpp:
```diff
@@ -48,8 +48,9 @@ void EditorSceneFormatImporterUFBX::get_extensions(List<String> *r_extensions) c
-	if (p_options.has("fbx/importer") && int(p_options["fbx/importer"]) == FBX_IMPORTER_FBX2GLTF) {
-		Ref<EditorSceneFormatImporterFBX> fbx2gltf_importer;
+	// FIXME: Hack to work around GH-86309.
+	if (p_options.has("fbx/importer") && int(p_options["fbx/importer"]) == FBX_IMPORTER_FBX2GLTF && GLOBAL_GET("filesystem/import/fbx2gltf/enabled")) {
```
Added a check for whether the fbx2gltf importer is actually enabled. Needs testing.

fire commented:
> The proper architecture should be scene/3d/model_*.h stuff as the abstraction for GLTF and FBX models. These CANNOT depend on any code in the gltf and fbx modules! Otherwise they're not abstractions at all. And that's a major problem currently so I had to move scene/3d/model_*.h back to the gltf module, after removing its coupling to the fbx module...

> If that's not possible, then the second best would be to have the fbx module depend fully on gltf (which it does in the current PR, despite the attempt at abstracting this away), reusing as much code as possible, instead of duplicating it like the original version of this PR did.

1. `scene/3d/model_*.h`'s abstraction is not used by anything other than me forcing fbx and gltf into a base class. @lyuma proposed deleting it.
2. fbx module depend fully on gltf document in practice
3. fbx extends gltf seems to fit more closely the code architecture.
4. We were thinking about deleting `scene/3d/model_*.h` as a pull request on top of this one because of the compatibility breakage on GLTFDocument.

@lyuma discussed with me removing the `scene/3d/model_*.h` and that seems like it would lead to a design where fbx module depend fully on gltf document, we discussed yesterday about this design and it has some merits. 


Thoughts?

akien-mga commented:
> @lyuma discussed with me removing the `scene/3d/model_*.h` and that seems like it would lead to a design where fbx module depend fully on gltf document, we discussed yesterday about this design and it has some merits.

Yeah I think it's the way to go if it's not doable to have a gltf/fbx agnostic abstraction :+1:

fire commented:
More commentary:

1. In the original version of the pr, the code was unfortunately duplicated, but the importers were independent.
2. In the second version of the pr, we removed the duplication but fbx and gltf documents became heavily coupled despite my attempt to extract the skinning common code and other common code.
3. We want less duplicated code, it means the coupling increases.
4. Given GLTFDocument and FBXDocument are so coupled we could probably extract common code, but I think the most minimal change is modify the object hierarchy so that FBX extends GLTF.
5. The tricky part is fbx would inherit some of the gltf api's extraneous methods relating to export, gltf extensions and the FBXState extending GLTFState. Perhaps there's a better design.

fire commented:
We've been testing using the fbx: https://sketchfab.com/3d-models/shibahu-2e9dffd948f747668609a5a477daad86.

This Shibahu model is fairly complicated and has blend shape animations which broke previously.

I'll compile and test.

akien-mga reviewed (approved):
Should be good to merge now.

We finally settled down on an architecture where the fbx module extends and depends on the gltf module, reusing what it can.

akien-mga commented:
Thanks! Amazing work @fire and @bqqbarbhg!

adamscott commented:
> Thanks! Amazing work @fire and @bqqbarbhg!

Yes indeed! Thanks both of you for your great job.

> Yeah I think it's the way to go if it's not doable to have a gltf/fbx agnostic abstraction 👍

Maybe I don't understand something, but in the face of it, it will make it possible to test that a FBX document is a GLTF document. This seems wrong. Maybe it's super difficult to implement an agnostic abstraction, but it still seems wrong.
